src/node_contextify.cc

```diff
@@ -929,6 +930,129 @@ class ContextifyScript : public BaseObject {
 };
 
 
+void ContextifyContext::GetWrappedFunction(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Isolate* isolate = env->isolate();
+  Local<Context> context = env->context();
+
+  // Argument 1: source code
+  CHECK(args[0]->IsString());
+  Local<String> code = args[0].As<String>();
+
+  // Argument 2: filename
+  CHECK(args[1]->IsString());
+  Local<String> filename = args[1].As<String>();
+
+  // Argument 3: line offset
+  CHECK(args[2]->IsNumber());
+  Local<Integer> line_offset = args[2].As<Integer>();
+
+  // Argument 4: column offset
+  CHECK(args[3]->IsNumber());
+  Local<Integer> column_offset = args[3].As<Integer>();
+
+  // Argument 5: cached data (optional)
+  Local<Uint8Array> cached_data_buf;
+  if (!args[4]->IsUndefined()) {
+    CHECK(args[4]->IsUint8Array());
+    cached_data_buf = args[4].As<Uint8Array>();
+  }
+
+  // Argument 6: produce cache data
+  CHECK(args[5]->IsBoolean());
+  bool produce_cached_data = args[5]->IsTrue();
+
+  // Argument 7: parsing context (optional)
+  Local<Context> parsing_context;
+  if (!args[6]->IsUndefined()) {
+    CHECK(args[6]->IsObject());
+    ContextifyContext* sandbox =
+      ContextifyContext::ContextFromContextifiedSandbox(
+          env, args[6].As<Object>());
+    CHECK_NOT_NULL(sandbox);
+    parsing_context = sandbox->context();
+  } else {
+    parsing_context = context;
+  }
+
+  // Argument 8: params for the function (optional)
+  Local<Array> params_buf;
+  if (!args[7]->IsUndefined()) {
+    CHECK(args[7]->IsArray());
+    params_buf = args[7].As<Array>();
+  }
+
+  // Read cache from cached data buffer
+  ScriptCompiler::CachedData* cached_data = nullptr;
+  if (!cached_data_buf.IsEmpty()) {
+    ArrayBuffer::Contents contents = cached_data_buf->Buffer()->GetContents();
+    uint8_t* data = static_cast<uint8_t*>(contents.Data());
+    cached_data = new ScriptCompiler::CachedData(
+      data + cached_data_buf->ByteOffset(), cached_data_buf->ByteLength());
+  }
+
+  ScriptOrigin origin(filename, line_offset, column_offset);
+  ScriptCompiler::Source source(code, origin, cached_data);
+  ScriptCompiler::CompileOptions options;
+  if (source.GetCachedData() == nullptr) {
+    options = ScriptCompiler::kNoCompileOptions;
+  } else {
+    options = ScriptCompiler::kConsumeCodeCache;
+  }
+
+  TryCatch try_catch(isolate);
+  Context::Scope scope(parsing_context);
+
+  // Read params from params buffer
+  size_t params_len;
+  Local<String>* params;
+  if (!params_buf.IsEmpty()) {
+    params_len = params_buf->Length();
+    params = new Local<String>[params_len];
+
+    for (int32_t n = 0; n < params_len; n++) {
+      Local<Value> val = params_buf->Get(context, n).ToLocalChecked();
+      CHECK(val->IsString());
```

Comment:
Sorry, Github cut off review here because my connection was flaky :/

For one, we don’t currently do any checking that `val` is actually a string in JS land, so right now passing in an array with non-string elements would crash the process (as far as I can tell)

Then, `params_buf->Get(…)` can fail if the array element is defined via a getter, so the `.ToLocalChecked()` may crash too – you probably want something like `if (!params_buf->Get(context, n).ToLocal(&val)) return;` (but just as a heads up, that would leak memory unless you go with an RAII option for managing `params` like the ones suggested above)

Reply:
Yes, this had been troubling me too, and the check was quite different originally. Will do this once I make the switch to vectors.

Reply:
i'd rather we did the argument handling in jsland and passed just raw arguments like ContextifyScript::New

Reply:
@devsnek totally doable. Plus then we could make it `throw` descriptive errors instead of just crashing.